pkg/mei/mkhiclient.go

```diff
@@ -0,0 +1,300 @@
+	}
+	// fw_sku is in bits 5-7 . ME_HFS3_FW_SKU_CUSTOM is 0x5
+	if (cs>>4)&0x7 == 0x5 {
+		// TODO implement the same as coreboot's cse_is_hfs1_com_soft_temp_disable()
+		return false, errors.New("IsHmrfpoEnableAllowed does not support checking for Soft Temp Disable yet")
+	}
+	return true, nil
+}
+
+type hmrfpoEnableMsg struct {
+	header MKHIHdr
+	nonce  uint32
+}
+
+func (hem hmrfpoEnableMsg) ToBytes() []byte {
+	var buf []byte
+	buf = append(buf, hem.header[:]...)
+	var nonce [4]byte
+	binary.LittleEndian.PutUint32(nonce[:], hem.nonce)
+	return append(buf, nonce[:]...)
+}
+
+type hmrfpoEnableResponse struct {
+	Header   MKHIHdr
+	FctBase  uint32
+	FctLimit uint32
+	Status   uint8
+	reserved [3]byte
+}
+
+func hmrfpoEnableResponseFromBytes(b []byte) (*hmrfpoEnableResponse, error) {
+	var resp hmrfpoEnableResponse
+	minlen := len(resp.Header)
+	maxlen := minlen +
+		4 /* FctBase */ +
+		4 /* FctLimit */ +
+		1 /* Status */ +
+		3 /* reserved bytes */
+	if len(b) != minlen && len(b) != maxlen {
+		return nil, fmt.Errorf("size mismatch, want %d/%d bytes, got %d", minlen, maxlen, len(b))
+	}
+	copy(resp.Header[:], b[:4])
+	if len(b) == minlen {
+		// don't parse the rest, we got a partial response
+		return &resp, nil
+	}
+	// TODO this could use u-root's pkg/uio
+	resp.FctBase = binary.LittleEndian.Uint32(b[4:8])
+	resp.FctLimit = binary.LittleEndian.Uint32(b[8:12])
+	resp.Status = b[12]
+	return &resp, nil
+}
+
+// HmrfpoEnable enables the HMRFPO (Host ME Region Flash Protection Override) via CSE,
+// see cse_hmrfpo_enable at
+// https://github.com/coreboot/coreboot/blob/b8b8ec832360ada5a313f10938bb6cfc310a11eb/src/soc/intel/common/block/include/intelblocks/cse.h#L64
+func (m MKHIClient) HmrfpoEnable() error {
+
+	var hdr MKHIHdr
+	hdr.SetGroupID(MKHIGroupIDHmrfpo)
+	hdr.SetCommand(MKHIHmrfpoEnable)
+	canEnable, err := m.IsHmrfpoEnableAllowed()
+	if err != nil {
+		return fmt.Errorf("Enabling HMRFPO failed: %w", err)
+	}
+	if !canEnable {
+		return fmt.Errorf("Enabling HMRFPO is not allowed")
+	}
+	msg := hmrfpoEnableMsg{
+		header: hdr,
+		nonce:  0,
+	}
+	if _, err := m.MEI.Write(msg.ToBytes()); err != nil {
+		return fmt.Errorf("write to MEI failed: %w", err)
+	}
+	buf := make([]byte, m.clientProperties.MaxMsgLength())
+	n, err := m.MEI.Read(buf)
+	if err != nil {
+		return fmt.Errorf("read from MEI failed: %w", err)
+	}
+	resp, err := hmrfpoEnableResponseFromBytes(buf[:n])
+	if err != nil {
+		return fmt.Errorf("failed to parse HMRFPOEnableResponse: %w", err)
+	}
+	if resp.Header.Result() != 0 {
+		return fmt.Errorf("failed to enable HMRFPO, request result is 0x%02x, want 0x0", resp.Header.Result())
+	}
+	if resp.Status != 0 {
+		return fmt.Errorf("failed to enable HMRFPO, request status is 0x%02x, want 0x0", resp.Status)
+	}
+	return nil
+}
+
+// GetMeiPciDevice will return the MEI PCI device object after scanning the PCI
+// bus.
+func GetMeiPciDevice() (*ghw.PCIDevice, error) {
+	pci, err := ghw.PCI()
```

Comment:
pkg/pci, to avoid that long list of new dependencies

Reply:
will do

Reply:
done